propagators/aws/xray/Propagator.php

```diff
@@ -0,0 +1,151 @@
+    private const TRACE_ID_DELIMITER_INDEX_2 = 10;
+    private const TRACE_ID_TIMESTAMP_LENGTH = 8;
+    private const TRACE_ID_RANDOMHEX_LENGTH = 24;
+
+    private const PARENT_ID_KEY = 'Parent';
+    private const SPAN_ID_LENGTH = 16;
+    
+    private const SAMPLED_FLAG_KEY = 'Sampled';
+    private const IS_SAMPLED = '1';
+    private const NOT_SAMPLED = '0';
+    
+    /**
+     * Returns list of fields used by HTTPTextFormat
+     */
+    public static function fields(): array
+    {
+        return [self::AWSXRAY_TRACE_ID_HEADER];
+    }
+
+    /**
+     * Creates and injects the traceHeader
+     * X-Amzn-Trace-Id: Root={traceId};Parent={parentId};Sampled={samplingFlag}
+     * X-Amzn-Trace-Id: Root=1-5759e988-bd862e3fe1be46a994272793;Parent=53995c3f42cd8ad8;Sampled=1
+     */
+    public static function inject(API\SpanContext $context, &$carrier, API\PropagationSetter $setter): void
+    {
+        $traceId = $context->getTraceId();
+        $spanId = $context->getSpanId();
+
+        if (!self::isValidSpanId($spanId) || !self::isValidTraceId($traceId)) return;
+
+        $timestamp = substr($traceId, 0, self::TRACE_ID_TIMESTAMP_LENGTH);
+        $randomHex = substr($traceId, self::TRACE_ID_TIMESTAMP_LENGTH);
+        $samplingFlag = $context->isSampled() ? self::IS_SAMPLED : self::NOT_SAMPLED;
+
+        $traceHeader = 'Root=' . self::VERSION_NUMBER . '-' . $timestamp . "-" . $randomHex . ";Parent=" . $spanId. ";Sampled=" . $samplingFlag;
+        $setter->set($carrier, self::AWSXRAY_TRACE_ID_HEADER, $traceHeader);
+    }
+
+    /**
+     * Extracts the span context from the carrier to transfer to the next service
+     * This function will parse all parts of the header and validate that it is
+     * formatted properly to AWS X-ray standards
+     */
+    public static function extract($carrier, API\PropagationGetter $getter): API\SpanContext
+    {
+        //Extract the traceHeader using the getter from the carrier
+        $traceHeader = $getter->get($carrier, self::AWSXRAY_TRACE_ID_HEADER);
+        
+        //If it is null it creates a dummy header with INVALID_TRACE and INVALID_SPAN
+        // restore(string $traceId, string $spanId, bool $sampled = false, bool $isRemote = false, ?API\TraceState $traceState = null): SpanContext
+        if ($traceHeader === null || gettype($traceHeader) !== 'string') {
+            return SpanContext::getInvalid();
+        }
+
+        $headerComponents = explode(';', $traceHeader);
+        $parsedTraceId = SpanContext::INVALID_TRACE;
+        $parsedSpanId = SpanContext::INVALID_SPAN;
+        $isSampled = false;
+
+        foreach($headerComponents as $component) {
+            //the key value pair of a component of the traceHeader. 
+            //Example: key = 'Parent' value = '53995c3f42cd8ad8'
+            $componentPair = explode('=', $component);
+
+            if ($componentPair[0] === 'Parent' && self::isValidSpanId($componentPair[1])) {
```

Comment:
Need to validate traceid, parentid and sampling decision. Once any of them is found invalid, should abort parsing and return invalid context. Need to adjust the unit tests as well.